projects/scannet_offline_eval/demo/proto_robot/mock_spot.py

```diff
@@ -0,0 +1,168 @@
+
+convert_tensor = transforms.ToTensor()
+center_crop = transforms.CenterCrop((224, 224))
+test_img_tensor = center_crop(convert_tensor(test_img))
+
+
+def tensor_to_bytes(x):
+    buff = io.BytesIO()
+    torch.save(x, buff)
+    return buff.getvalue()
+
+
+def tensor_to_robotensor(x):
+    robo_tensor = robodata_pb2.RoboTensor()
+    robo_tensor.dtype = str(x.dtype)
+    tensorshape = x.shape
+    dim_fields = [robo_tensor.d1, robo_tensor.d2, robo_tensor.d3]
+    if len(tensorshape) <= 3:
+        for i in range(len(tensorshape)):
+            dim_fields[i] = tensorshape[i]
+    robo_tensor.tensor_content = tensor_to_bytes(x)
+    return robo_tensor
+
+
+def create_msg():
+    new_msg = robodata_pb2.RobotSummary()
+    new_msg.message = "testing robot data"
+    yield new_msg
+
+
+def create_msg_from_file(data, idx):
+    print("generating message for " + str(idx))
+    new_msg = robodata_pb2.RobotSummary()
+
+    new_msg.rgb_tensor.CopyFrom(tensor_to_robotensor(data["rgb"][idx]))
+    new_msg.depth_tensor.CopyFrom(tensor_to_robotensor(data["depth"][idx]))
+
+    new_msg.message = "testing robot data"
+    new_msg.robot_obs.gps.lat = data["obs"][idx].gps[0]
+    new_msg.robot_obs.gps.long = data["obs"][idx].gps[1]
+
+    yield new_msg
+
+
+def send_robot_data(stub):
+    for d in range(len(data) - 1):
+        new_msg = create_msg_from_file(data, d)
+        print(new_msg)
+        responses = stub.ReceiveRobotData(new_msg)
+        for response in responses:
+            print("Received message %s " % (response.message))
+
+
+def read_robot_data(stub):
+    for i in range(4):
+        response = stub.GetHistory(robodata_pb2.RobotSummary())
+        print("Getting History...")
+        for r in response:
+            print(torch.load(io.BytesIO(r.rgb_tensor.tensor_content)))
+
+
+def test_conversation(stub):
+    new_conv = robodata_pb2.LLMInput()
+    chat = new_conv.conversation.add()
+    chat.role = "User"
+    chat.content = "Hey! Whats up"
+    img = new_conv.imgs.add()
+    img.CopyFrom(tensor_to_robotensor(test_img_tensor))
+    stub.Chat(new_conv)
+
+
+def test_vlm(stub):
+    voxel_file = "/private/home/ssilwal/dev/home-robot/spot_output.pkl"
+    voxel_map = v.SparseVoxelMap()
+    voxel_map.read_from_pickle(voxel_file)
+    print("generated voxel_map")
+
+    def get_obj_centric_world_representation(instance_memory, max_context_length):
+        crops = []
+        for global_id, instance in enumerate(instance_memory):
+            instance_crops = instance.instance_views
+            crops.append((global_id, random.sample(instance_crops, 1)[0].cropped_image))
+        # TODO: the model currenly can only handle 20 crops
+        if len(crops) > max_context_length:
+            print(
+                "\nWarning: this version of minigpt4 can only handle limited size of crops -- sampling a subset of crops from the instance memory..."
+            )
+            crops = random.sample(crops, max_context_length)
+        import shutil
```

Comment:
Don't import in the middle of a function like this

Reply:
removed